### ISSUE TITLE
Combining coefficients

### ISSUE THREAD

**Issue by macartan** [DeclareDesign/DeclareDesign#178]
@acoppock 
[Issue perhaps for estimatr] 
Often we are interested in various combinations of coefficients; eg main effect plus interaction term, or average effect across two conditions in factorial, or difference between two coefficients in m arm trial;

Example: Ideal syntax might be sthg like (three arm trial):

declare_estimator(
  Y ~ X + Z,
  coefficients = c("Z1", "Z2", "Z2-Z1")
  model = estimatr::lm_robust
  estimand = c("E1", "E2", "E3")
)

or

declare_estimator(
  Y ~ X + Z,
  coefficients = c("Z1",  "Z1+Z1:Z2",  "Z1+.5*Z1:Z2", )
  ....

for factorial; note the ".5"; perhaps for the average effect of Z1 across conditions

More generally perhaps:
 
declare_estimator(
  Y ~ X + Z,
  coefficients = c("Z1",  f(coefficients)),
   ...

**Comment by nfultz** [comment]
Generally I would do this using `car::linear.hypothesis` - since this is a
third party package we should probably put this feature in a new handler
(in designs maybe? or estimatr) and not in the core package. It should be
possible since estimatr already supports coef() and vcov().

**Comment by lilymedina** [comment]
just reopening this issue.

From a conversation today, we think that this feature could be included as argument `combine_coefficients` in `declare_estimator()` so that it'd look like this:
 
```
declare_estimator(
Y ~ Z1  + Z2,
term= c("Z1", "Z2")
combine_coefficients = c("Z2-Z1")
model = lm_robust
estimand = c("E1", "E2", "E3")
)
```
As Neal suggested estimation can be done using `car::linearHypothesis`.


Now, I'm wondering if this should be a function sitting in `R/declare_estimator.R ` and that gets evaluated once the model has been estimated if `!is.NULL(combine_coefficients)`.



@graemeblair Macartan mentioned that you've given it some thoughts to this and that you might have an idea on how to implement.

**Comment by nfultz** [comment]
Probably not in declare_estimator.R, that is for capturing arguments and integrating them into a design steps objects, it does not do any actual estimation or calculation itself, it only passes it's arguments through to the model function. This probably belongs as a feature in estimatr::lm_robust or by writing a new type of step.

**Comment by lilymedina** [comment]
I see. So it makes more sense that it sits in `estimatr `. The only thing is that it would be good to be able to use it for all model_handlers.

**Comment by macartan** [comment]
I think that makes sense, car:linear hypothesis works for many models so
would it be nice for an estimation step to (1) pass to model function (2)
get tidy output back but also coef/vcov back when possible & required (3)
send these to a function that uses linear hypothesis (4) returns a
estimator/estimand table that looks just like the current output so as not
to alter other parts of pipe?

**Comment by nfultz** [comment]
I drafted a vignette for dealing with contrasts, linear combs, and multiple testing adjustments - https://github.com/DeclareDesign/DeclareDesign/blob/27f0e405dbd2ee5cb92216f6e7330ace0130f210/vignettes/postestimation.Rmd

The bad (but unsurprising) news is there is basically no consistent way in R to do things stata would call "postestimation"

Also many useful packages aren't yet supported by tidy :(

The good news is that wrapping their bespoke apis in a tidy way is reasonably straightforward albeit verbose. 

One thing we should think about is whether it is desirable to have different types of things (with different units) in a single output column, eg both t-stats and f-stats in "statistic", or both coefficients and eta-squareds in "estimate" for example.

**Comment by lilymedina** [comment]
Implementing this feature by creating function [lh_robust](https://github.com/DeclareDesign/estimatr/blob/combine_coefficients/R/estimatr_lh_robust.R)  as @graemeblair suggested on a different thread. 

`Tidy() ` would print model and linear combination estimates, se, t etc... (see below)

One limitation with this option is that it can only be used with lm_robust, whereas car::linearHypothesis can work with many more.

thoughts? 


```
lmh <- lh_robust(cyl ~ hp + am, data = mtcars,  linearHypothesis = "hp + am = 0", lh_term = "lh: hp+am=0")
```
This is how tidy.lh_robust prints
`tidy(lmh)`
<table>
<thead>
<tr class="header">
<th></th>
<th align="left">term</th>
<th align="right">estimate</th>
<th align="right">std.error</th>
<th align="right">statistic</th>
<th align="right">p.value</th>
<th align="right">conf.low</th>
<th align="right">conf.high</th>
<th align="right">df</th>
<th align="left">outcome</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>lm_robust.1</td>
<td align="left">(Intercept)</td>
<td align="right">3.8177680</td>
<td align="right">0.5370020</td>
<td align="right">7.109411</td>
<td align="right">0.0000001</td>
<td align="right">2.7194756</td>
<td align="right">4.9160604</td>
<td align="right">29</td>
<td align="left">cyl</td>
</tr>
<tr class="even">
<td>lm_robust.2</td>
<td align="left">hp</td>
<td align="right">0.0195279</td>
<td align="right">0.0027532</td>
<td align="right">7.092920</td>
<td align="right">0.0000001</td>
<td align="right">0.0138971</td>
<td align="right">0.0251587</td>
<td align="right">29</td>
<td align="left">cyl</td>
</tr>
<tr class="odd">
<td>lm_robust.3</td>
<td align="left">am</td>
<td align="right">-1.2178820</td>
<td align="right">0.3423388</td>
<td align="right">-3.557534</td>
<td align="right">0.0013100</td>
<td align="right">-1.9180434</td>
<td align="right">-0.5177205</td>
<td align="right">29</td>
<td align="left">cyl</td>
</tr>
<tr class="even">
<td>linearHypothesis</td>
<td align="left">lh: hp+am=0</td>
<td align="right">-1.1983541</td>
<td align="right">0.3441508</td>
<td align="right">3.482061</td>
<td align="right">1.9984020</td>
<td align="right">-1.9022216</td>
<td align="right">-0.4944866</td>
<td align="right">29</td>
<td align="left">cyl</td>
</tr>
</tbody>
</table>


And other functions...

```
print(lmh)

    $lm_robust
                   Estimate  Std. Error   t value     Pr(>|t|)    CI Lower
    (Intercept)  3.81776799 0.537001992  7.109411 8.011871e-08  2.71947560
    hp           0.01952788 0.002753152  7.092920 8.368300e-08  0.01389706
    am          -1.21788195 0.342338809 -3.557534 1.310033e-03 -1.91804343
                   CI Upper DF
    (Intercept)  4.91606039 29
    hp           0.02515871 29
    am          -0.51772047 29

    $linearHypothesis
                 Estimate Std. Error  t value Pr(>|t|)  CI Lower   CI Upper DF
    lh: hp+am=0 -1.198354  0.3441508 3.482061 1.998402 -1.902222 -0.4944866 29


summary(lmh)

    $lm_robust 
     
    Call:
    NULL

    Standard error type:  HC2 

    Coefficients:
                Estimate Std. Error t value  Pr(>|t|) CI Lower CI Upper DF
    (Intercept)  3.81777   0.537002   7.109 8.012e-08   2.7195  4.91606 29
    hp           0.01953   0.002753   7.093 8.368e-08   0.0139  0.02516 29
    am          -1.21788   0.342339  -3.558 1.310e-03  -1.9180 -0.51772 29

    Multiple R-squared:  0.8019 ,   Adjusted R-squared:  0.7882 
    F-statistic:  84.3 on 2 and 29 DF,  p-value: 8.244e-13


    $linearHypothesis 
     
                 Estimate Std. Error  t value Pr(>|t|)  CI Lower   CI Upper DF
    lh: hp+am=0 -1.198354  0.3441508 3.482061 1.998402 -1.902222 -0.4944866 29

    Linear hypothesis test

    Hypothesis:
    hp  + am = 0

    Model 1: restricted model
    Model 2: cyl ~ hp + am

      Res.Df Df  Chisq Pr(>Chisq)    
    1     30                         
    2     29  1 12.125  0.0004976 ***
    ---
    Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

```

**Comment by graemeblair** [comment]
We confirmed that this plays nicely with DeclareDesign.  You have to type the name of the term 2x which is annoying, but not sure how to fix.  Thank you! (small edits on the estimatr PR)

```{r}
library(DeclareDesign)


design <- 
  declare_population(N = 100, u = rnorm(N), X = rbinom(N, 1, 0.5)) +
  declare_potential_outcomes(Y ~ 3 * Z + 2 * Z * X + u) +
  declare_estimand(cate = mean(Y_Z_1 - Y_Z_0), subset = X == 1) +
  declare_assignment(m = 50) +
  declare_reveal(Y, Z) +
  declare_estimator(Y ~ Z*X, model = lh_robust, linearHypothesis = "Z + Z:X = 0", term = "Z + Z:X = 0", estimand = "cate")

dx <- diagnose_design(design, sims = 100, bootstrap_sims = FALSE)

dx

```

**Comment by nfultz** [comment]
If you had a long or lot of linear hypotheses, I think you can do something like:

```
my_hyp = "Z + Z:X = 0"

e <- declare_estimator(Y ~ Z*X, model = lh_robust, linearHypothesis = !!my_hyp, term = !!my_hyp, estimand = "cate")

```

**Comment by lilymedina** [comment]
implemented in [#281](https://github.com/DeclareDesign/estimatr/pull/281)

**Comment by nfultz** [comment]
> You have to type the name of the term 2x which is annoying, but not sure how to fix.

IIRC, the current defaults for term is something like `tidy(model)[-1,]` - with the intent of stripping out the intercept. We may want to make it less braindead (eg `filter(tidy(model), term != '(Intercept)'`).
